File: daily-exercise/high_frequency/valid-parentheses.py
# -*- coding: utf-8 -*-
# @Time    : 2022/7/27 20:58
# @Author  : kevin
# @Version : python 3.7
# @Desc    : 有效的括号

# 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。
# ()[]{} {()} (] ([)] ){  )) (())
# ( (
class Solution:
    def isValid(self, s):
        if len(s) % 2 == 1:  # 字符串长度必须为偶数
            return False

        stack = []  # 栈结构
        parentheses = {'(': ')', '{': '}', '[': ']'}  # 括号关系map
        for item in s:
            if item in parentheses.keys():  # 如果相同方向括号必须加进来
                stack.append(parentheses[item])
            else:
                if not stack or stack[-1] != item:  # 栈为空（都闭合了）首个括号就逆向 或者 最新加入也就是栈顶的括号不等于当前元素
                    return False
                else:  # 闭合一个之后立即弹出栈顶括号
                    stack.pop()
        return True if not stack else False  # 两边重复括号完全一致(())/()()


s = Solution()
bo = s.isValid("()()")

File: daily-exercise/high_frequency/valid-parentheses.go
// byte表示ASCII字节，go中uint8类型；rune表示Unicode字符（包含中文），go中int32类型，占4个byte
// go 字符串循环，不输出中文及byte形式输出，用下标索引；输出中文，用 for range

func isValid(s string) bool {
    n := len(s)
    if n %2 == 1{
        return false
    }

    stack := make([]byte,0)
    parentheses := map[byte]byte{'(':')', '[':']', '{':'}',}
    for i := 0; i < n; i++  {
        if val,ok := parentheses[s[i]];ok {
            stack = append(stack,val)
        } else {
            if len(stack) == 0 || stack[len(stack)-1] != s[i] {
                return false
            } else {
                stack = stack[:len(stack)-1]
            }
        }
    }
    if len(stack) == 0 {
        return true
    } else {
        return false
    }
}